pkg/internal/resources/apiserver/apiserver.go

```diff
@@ -48,133 +48,194 @@ var k = kustomize.NewDefaultKustomize()
 	v := version.Get()
 	kc := k.ForHTTP(vfs)
-	if err := kc.MkLayer("man", types.Kustomization{
-		Namespace: c.Namespace,
-		Images: []image.Image{
-			{
-				Name:   "quay.io/kubecarrier/apiserver",
-				NewTag: v.Version,
-			},
-		},
-		PatchesStrategicMerge: []types.PatchStrategicMerge{
-			"manager_env_patch.yaml",
-		},
-		Resources: []string{"../default"},
-	}); err != nil {
-		return nil, fmt.Errorf("cannot mkdir: %w", err)
-	}
-
-	extraArgs := make([]string, 0)
-	if len(c.Spec.OIDC.RequiredClaims) > 0 {
-		rclaims := make([]string, 0)
-		for k, v := range c.Spec.OIDC.RequiredClaims {
-			rclaims = append(rclaims, fmt.Sprintf("%s=%s", k, v))
+	resources := []string{"../default"}
+	if c.Spec.TLSSecretRef == nil || c.Spec.TLSSecretRef.Name == "" {
+		// TLS Secret is not present, we create a self-signed certificate for localhost via cert-manager.
+		resources = append(resources, "../certmanager")
+		c.Spec.TLSSecretRef = &operatorv1alpha1.ObjectReference{
+			Name: "apiserver-tls-cert",
 		}
-		extraArgs = append(extraArgs, "--oidc-required-claim="+strings.Join(rclaims, ","))
 	}
-
 	// Patch environment
 	// Note:
 	// we are not using *appsv1.Deployment here,
 	// because some fields will be defaulted to empty and
 	// interfere with the strategic merge patch of kustomize.
-	managerEnv := map[string]interface{}{
-		"apiVersion": "apps/v1",
-		"kind":       "Deployment",
-		"metadata": map[string]string{
-			"name":      "manager",
-			"namespace": "system",
-		},
-		"spec": map[string]interface{}{
-			"template": map[string]interface{}{
-				"spec": map[string]interface{}{
-					"containers": []map[string]interface{}{
-						{
-							"name": "manager",
-							"args": append([]string{
-								"--address=$(API_SERVER_ADDR)",
-								"--tls-cert-file=$(API_SERVER_TLS_CERT_FILE)",
-								"--tls-private-key-file=$(API_SERVER_TLS_PRIVATE_KEY_FILE)",
-								"--oidc-issuer-url=$(API_SERVER_OIDC_ISSUER_URL)",
-								"--oidc-client-id=$(API_SERVER_OIDC_CLIENT_ID)",
-								"--oidc-ca-file=$(API_SERVER_OIDC_CA_FILE)",
-								"--oidc-username-claim=$(API_SERVER_OIDC_USERNAME_CLAIM)",
-								"--oidc-username-prefix=$(API_SERVER_OIDC_USERNAME_PREFIX)",
-								"--oidc-groups-claim=$(API_SERVER_OIDC_GROUPS_CLAIM)",
-								"--oidc-groups-prefix=$(API_SERVER_OIDC_GROUPS_PREFIX)",
-								"--oidc-signing-algs=$(API_SERVER_OIDC_SIGNING_ALGS)",
-							}, extraArgs...),
-							"env": []map[string]interface{}{
-								{
-									"name":  "API_SERVER_ADDR",
-									"value": ":8443",
-								},
-								{
-									"name":  "API_SERVER_TLS_CERT_FILE",
-									"value": "/run/serving-certs/tls.crt",
-								},
-								{
-									"name":  "API_SERVER_TLS_PRIVATE_KEY_FILE",
-									"value": "/run/serving-certs/tls.key",
-								},
-								{
-									"name":  "API_SERVER_OIDC_ISSUER_URL",
-									"value": c.Spec.OIDC.IssuerURL,
+	var managerEnv map[string]interface{}
+	if c.Spec.OIDC != nil {
```

Comment:
I also wanted to use `Deployment` to deduplicate this, just remembered that we had issues with it(some fields that we did't want to). @thetechnick Do you remember what was the issue?

Reply:
I have tried to swtich to using `Deployment`, and I looked into our golden file, only `deployment.Spec.Strategy` and `deployment.Status` are filled with `{}` empty values, so I think it should be fine? and if we still want to remove these empty values, we can still convert `Deployment` to `Unstructured` and remove these keys before marshaling.

Also, I would like to use `Deployment` rather than `map[string]interface{}`, because something can go wrong easily with map, for example, the key in our `VolumeMounts` https://github.com/kubermatic/kubecarrier/blob/feature-api-server/pkg/internal/resources/apiserver/apiserver.go#L155
It should be `readOnly` not `readyOnly`

Reply:
yes yes yes! on not using `map[string]interface{}` 

hmm, which serializer/encoded have you used? for `.status` I'm not at all worried about since this wouldn't be applied to the objects due to `/status` subresource.

for `.spec.strategy` would the Kubernetes API correctly apply the defaults here? If so...than let's roll with it!

Reply:
actually I am just using `yaml.Marshal()` and it works :)